components/apimgt/org.wso2.carbon.apimgt.rest.api.util/src/main/java/org/wso2/carbon/apimgt/rest/api/util/impl/OAuthJwtAuthenticatorImpl.java

```diff
@@ -0,0 +1,347 @@
+                        }
+                    }
+                    if (log.isDebugEnabled()) {
+                        log.debug("username = " + username);
+                    }
+                    tenantId = realmService.getTenantManager().getTenantId(tenantDomain);
+                    carbonContext.setTenantDomain(tenantDomain);
+                    carbonContext.setTenantId(tenantId);
+                    carbonContext.setUsername(username);
+                    if (!tenantDomain.equals(MultitenantConstants.SUPER_TENANT_DOMAIN_NAME)) {
+                        APIUtil.loadTenantConfigBlockingMode(tenantDomain);
+                    }
+                    return true;
+                } catch (UserStoreException e) {
+                    log.error("Error while retrieving tenant id for tenant domain: " + tenantDomain, e);
+                }
+                log.info("Scope validation success");
+                return true;
+            }
+            log.error("scopes validation failed");
+            return false;
+        }
+        log.error("scopes validation failed");
+        return false;
+    }
+
+    /**
+     * Get signed jwt info.
+     *
+     * @param accessToken JWT token
+     * @return SignedJWTInfo : Signed token info
+     */
+    private SignedJWTInfo getSignedJwt(String accessToken) throws ParseException {
+
+        SignedJWT signedJWT = SignedJWT.parse(accessToken);
+        JWTClaimsSet jwtClaimsSet = signedJWT.getJWTClaimsSet();
+        return new SignedJWTInfo(accessToken, signedJWT, jwtClaimsSet);
+    }
+
+    /**
+     * Validate the JWT token.
+     *
+     * @param jti           jwtTokenIdentifier
+     * @param signedJWTInfo signed jwt info object
+     * @return JWTValidationInfo : token validated info
+     */
+    private JWTValidationInfo validateJWTToken(SignedJWTInfo signedJWTInfo, String jti, String accessToken,
+                                               String jwtHeader, URL basePath) throws APIManagementException {
+
+        JWTValidationInfo jwtValidationInfo;
+        String issuer = signedJWTInfo.getJwtClaimsSet().getIssuer();
+
+        if (StringUtils.isNotEmpty(issuer)) {
+            //validate Issuer
+            List<String> audiences = signedJWTInfo.getJwtClaimsSet().getAudience();
+            if (tokenIssuers != null && tokenIssuers.containsKey(issuer)) {
+                jwtValidator.loadTokenIssuerConfiguration(tokenIssuers.get(issuer));
+                //validate audience
+                if (audiencesList != null && audiences.contains(audiencesList.get(basePath.getPath()))) {
+                    if (isRESTApiTokenCacheEnabled) {
+                        JWTValidationInfo tempJWTValidationInfo = (JWTValidationInfo) getRESTAPITokenCache().get(jti);
+                        if (tempJWTValidationInfo != null) {
+                            Boolean isExpired = checkTokenExpiration(new Date(tempJWTValidationInfo.getExpiryTime()));
```

Comment:
Shouldn't we validate the expiration of token before doing audience validation?

Reply:
For the signature and expiry validation we are utilising a common method used for gateway JWT validation and both exp and signature validation parts handle by that common method `(jwtValidator.validateToken)`. So we brought forward possible validation parts (issuer and aud) by considering the performance impact. When retrieving the token from the cache we do expiration check and proceed.